addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone_test.go

```diff
@@ -299,8 +301,198 @@ var _ = Describe("Clusterbootstrap", func() {
+			clusterBootstrap := constructFakeEmptyClusterBootstrap()
+			clusterBootstrap.Spec = &v1alpha3.ClusterBootstrapTemplateSpec{
+				CNI: &v1alpha3.ClusterBootstrapPackage{
+					RefName: "calico*",
+					ValuesFrom: &v1alpha3.ValuesFrom{
+						Inline: map[string]interface{}{"foo": "bar"},
+					},
+				},
+			}
+			tanzuKubernetesRelease := constructFakeTanzuKubernetesRelease()
+			err := helper.CompleteCBPackageRefNamesFromTKR(tanzuKubernetesRelease, clusterBootstrap)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(clusterBootstrap.Spec.CNI.RefName).To(Equal("calico.tanzu.vmware.com.3.22.1+vmware.1-tkg.1-zshippable"))
+			Expect(clusterBootstrap.Spec.CNI.ValuesFrom.Inline["foo"]).To(Equal("bar"))
+		})
+
+		It("should return error if there is a no match", func() {
+			clusterBootstrap := constructFakeEmptyClusterBootstrap()
+			clusterBootstrap.Spec = &v1alpha3.ClusterBootstrapTemplateSpec{
+				CNI: &v1alpha3.ClusterBootstrapPackage{
+					RefName: "something-does-not-exist*",
+					ValuesFrom: &v1alpha3.ValuesFrom{
+						Inline: map[string]interface{}{"foo": "bar"},
+					},
+				},
+			}
+			tanzuKubernetesRelease := constructFakeTanzuKubernetesRelease()
+			err := helper.CompleteCBPackageRefNamesFromTKR(tanzuKubernetesRelease, clusterBootstrap)
+			Expect(err).To(HaveOccurred())
+			// The original value should stay untouched
+			Expect(clusterBootstrap.Spec.CNI.RefName).To(Equal("something-does-not-exist*"))
+		})
+
+		It("should not touch the fully filled refName", func() {
+			clusterBootstrap := constructFakeEmptyClusterBootstrap()
+			clusterBootstrap.Spec = &v1alpha3.ClusterBootstrapTemplateSpec{
+				CNI: &v1alpha3.ClusterBootstrapPackage{
+					RefName: "calico*",
+					ValuesFrom: &v1alpha3.ValuesFrom{
+						Inline: map[string]interface{}{"foo": "bar"},
+					},
+				},
+				CPI: &v1alpha3.ClusterBootstrapPackage{
+					RefName: "fake-cpi",
+					ValuesFrom: &v1alpha3.ValuesFrom{
+						ProviderRef: &corev1.TypedLocalObjectReference{
+							Name: "fake-secret",
+							Kind: "secret",
+						},
+					},
+				},
+			}
+			tanzuKubernetesRelease := constructFakeTanzuKubernetesRelease()
+			err := helper.CompleteCBPackageRefNamesFromTKR(tanzuKubernetesRelease, clusterBootstrap)
+			Expect(err).NotTo(HaveOccurred())
+			// The original value should stay untouched
+			Expect(clusterBootstrap.Spec.CPI.RefName).To(Equal("fake-cpi"))
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.Inline).To(BeNil())
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.SecretRef).To(BeEmpty())
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef.Kind).To(Equal("secret"))
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef.Name).To(Equal("fake-secret"))
+			// The partial filled refName should be updated
+			Expect(clusterBootstrap.Spec.CNI.RefName).To(Equal("calico.tanzu.vmware.com.3.22.1+vmware.1-tkg.1-zshippable"))
+			Expect(clusterBootstrap.Spec.CNI.ValuesFrom.Inline["foo"]).To(Equal("bar"))
+		})
+	})
```

Comment:
Should we add a test for merge of additionalPackages? Request has one additionalPackage and the rest are added from CBT? We should ensure that more than 1 package with same refName doesn't exist. Example user provides fluentbit.tanzu.vmware.com.3.1 whereas CBT has fluentbit.tanzu.vmware.com.2.1

Reply:
Done, added more tests with non-empty `AdditionalPackages`